From FEDORA_PATCHES Mon Sep 17 00:00:00 2001
From: Fedora GDB patches <invalid@email.com>
Date: Fri, 27 Oct 2017 21:07:50 +0200
Subject: gdb-glibc-strstr-workaround.patch

;; Workaround PR libc/14166 for inferior calls of strstr.
;;=fedoratest: Compatibility with RHELs (unchecked which ones).

diff --git a/gdb/testsuite/gdb.base/gnu-ifunc-strstr-workaround.exp b/gdb/testsuite/gdb.base/gnu-ifunc-strstr-workaround.exp
new file mode 100644
--- /dev/null
+++ b/gdb/testsuite/gdb.base/gnu-ifunc-strstr-workaround.exp
@@ -0,0 +1,119 @@
+# Copyright (C) 2012 Free Software Foundation, Inc.
+
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+# Workaround for:
+# invalid IFUNC DW_AT_linkage_name: memmove strstr time
+# http://sourceware.org/bugzilla/show_bug.cgi?id=14166
+
+if {[skip_shlib_tests]} {
+    return 0
+}
+
+set testfile "gnu-ifunc-strstr-workaround"
+set executable ${testfile}
+set srcfile start.c
+set binfile [standard_output_file ${executable}]
+
+if [prepare_for_testing ${testfile}.exp $executable $srcfile] {
+    return -1
+}
+
+if ![runto_main] {
+    return 0
+}
+
+set test "ptype atoi"
+gdb_test_multiple $test $test {
+    -re "type = int \\(const char \\*\\)\r\n$gdb_prompt $" {
+	pass $test
+    }
+    -re "type = int \\(\\)\r\n$gdb_prompt $" {
+	untested "$test (no DWARF)"
+	return 0
+    }
+    -re "type = <unknown return type> \\(\\)\r\n$gdb_prompt $" {
+	untested "$test (no DWARF)"
+	return 0
+    }
+}
+
+set addr ""
+set test "print strstr"
+gdb_test_multiple $test $test {
+    -re " = {<text gnu-indirect-function variable, no debug info>} (0x\[0-9a-f\]+) <strstr>\r\n$gdb_prompt $" {
+	set addr $expect_out(1,string)
+	pass $test
+    }
+    -re " = {<text gnu-indirect-function variable, no debug info>} (0x\[0-9a-f\]+) <__strstr>\r\n$gdb_prompt $" {
+	set addr $expect_out(1,string)
+	pass "$test (GDB workaround)"
+    }
+    -re " = {<text gnu-indirect-function variable, no debug info>} (0x\[0-9a-f\]+) <__libc_strstr>\r\n$gdb_prompt $" {
+	set addr $expect_out(1,string)
+	pass "$test (fixed glibc)"
+    }
+    -re " = {<text gnu-indirect-function variable, no debug info>} (0x\[0-9a-f\]+) <__libc_strstr_ifunc>\r\n$gdb_prompt $" {
+	set addr $expect_out(1,string)
+	pass "$test (fixed glibc)"
+    }
+    -re " = {char \\*\\(const char \\*, const char \\*\\)} 0x\[0-9a-f\]+ <strstr>\r\n$gdb_prompt $" {
+	untested "$test (gnu-ifunc not in use by glibc)"
+	return 0
+    }
+}
+
+set test "info sym"
+gdb_test_multiple "info sym $addr" $test {
+    -re "strstr in section \\.text of /lib\[^/\]*/libc.so.6\r\n$gdb_prompt $" {
+	pass $test
+    }
+    -re " = {char \\*\\(const char \\*, const char \\*\\)} 0x\[0-9a-f\]+ <strstr>\r\n$gdb_prompt $" {
+	# unexpected
+	xfail "$test (not in libc.so.6)"
+	return 0
+    }
+}
+
+set test "info addr strstr"
+gdb_test_multiple $test $test {
+    -re "Symbol \"strstr\" is a function at address $addr\\.\r\n$gdb_prompt $" {
+	fail "$test (DWARF for strstr)"
+    }
+    -re "Symbol \"strstr\" is at $addr in a file compiled without debugging\\.\r\n$gdb_prompt $" {
+	pass "$test"
+    }
+}
+
+set test "print strstr second time"
+gdb_test_multiple "print strstr" $test {
+    -re " = {<text gnu-indirect-function variable, no debug info>} $addr <strstr>\r\n$gdb_prompt $" {
+	pass $test
+    }
+    -re " = {<text gnu-indirect-function variable, no debug info>} $addr <__strstr>\r\n$gdb_prompt $" {
+	pass "$test (GDB workaround)"
+    }
+    -re " = {<text gnu-indirect-function variable, no debug info>} $addr <__libc_strstr>\r\n$gdb_prompt $" {
+	pass "$test (fixed glibc)"
+    }
+    -re " = {<text gnu-indirect-function variable, no debug info>} $addr <__libc_strstr_ifunc>\r\n$gdb_prompt $" {
+	pass "$test (fixed glibc)"
+    }
+    -re " = {void \\*\\(void\\)} 0x\[0-9a-f\]+ <strstr>\r\n$gdb_prompt $" {
+	fail $test
+    }
+}
+
+gdb_test {print (char *)strstr("abc","b")} { = 0x[0-9a-f]+ "bc"}
+gdb_test {print (char *)strstr("def","e")} { = 0x[0-9a-f]+ "ef"}
